test/integration/concurrent_users_test.go

```diff
@@ -0,0 +1,234 @@
+
+	var usersNames = make([]string, 0, numUsers)
+
+	for i := range numUsers {
+		user, err := testutil.NewUser(
+			&token.User{
+				ID:       fmt.Sprintf("user-%d", i+1),
+				Username: fmt.Sprintf("user-%d@acme.com", i+1),
+				Groups:   []string{},
+			},
+			gatewayPort,
+			kindKubeConfig.Host,
+			controller.URL,
+		)
+		require.NoError(t, err, "failed to create user %d", i+1)
+
+		users = append(users, user)
+		usersNames = append(usersNames, user.Username)
+	}
+
+	defer func() {
+		for _, user := range users {
+			user.Close()
+		}
+	}()
+
+	testutil.CreateK8sRoleBinding(t, kindKubectl, usersNames)
+
+	wg := sync.WaitGroup{}
+
+	for _, user := range users {
+		wg.Add(1)
+
+		go func() {
+			expectedUser := map[string]any{
+				"id":       user.ID,
+				"username": user.Username,
+				"groups":   []any{},
+			}
+
+			commands := []command{
+				{
+					name: "whoami",
+					args: []string{"auth", "whoami", "-o", "json"},
+					assertOutputFn: func(t *testing.T, output []byte) {
+						t.Helper()
+						testutil.AssertWhoAmI(t, output, user.Username, []string{"system:authenticated"})
+					},
+					assertLogFn: func(t *testing.T, logs *observer.ObservedLogs) {
+						t.Helper()
+						testutil.AssertLogsForREST(t, logs, "/apis/authentication.k8s.io/v1/selfsubjectreviews", expectedUser, http.StatusCreated)
+					},
+				},
+				{
+					name:           "get-pods",
+					args:           []string{"get", "pods", "-o", "json"},
+					assertOutputFn: testutil.AssertGetPods,
+					assertLogFn: func(t *testing.T, logs *observer.ObservedLogs) {
+						t.Helper()
+						testutil.AssertLogsForREST(t, logs, "/api/v1/namespaces/default/pods?limit=500", expectedUser, http.StatusOK)
+					},
+				},
+				{
+					name: "exec-sleep",
+					args: []string{"exec", "test-pod", "--", "sleep", "2"},
+					assertOutputFn: func(t *testing.T, output []byte) {
+						t.Helper()
+						assert.Empty(t, string(output))
+					},
+					assertLogFn: func(t *testing.T, logs *observer.ObservedLogs) {
+						t.Helper()
+						testutil.AssertLogsForExec(t, logs, "/api/v1/namespaces/default/pods/test-pod/exec?command=sleep&command=2&container=test-pod&stderr=true&stdout=true", "", expectedUser)
+					},
+				},
+				{
+					name: "exec-cat",
+					args: []string{"exec", "test-pod", "--", "cat", "/etc/hostname"},
+					assertOutputFn: func(t *testing.T, output []byte) {
+						t.Helper()
+						assert.Equal(t, "test-pod\n", string(output))
+					},
+					assertLogFn: func(t *testing.T, logs *observer.ObservedLogs) {
+						t.Helper()
+						testutil.AssertLogsForExec(t, logs, "/api/v1/namespaces/default/pods/test-pod/exec?command=cat&command=%2Fetc%2Fhostname&container=test-pod&stderr=true&stdout=true", "test-pod", expectedUser)
+					},
+				},
+			}
+			rand.Shuffle(len(commands), func(i, j int) {
+				commands[i], commands[j] = commands[j], commands[i]
+			})
+
+			t.Logf("User %s is running commands: %s", user.ID, commands)
+
+			lastLogTime := time.Time{}
```

Comment:
can this be inside the `for` loop, before running the command?

Reply:
Yes, I think it's nicer! We can remove [line 215](https://github.com/Twingate/kubernetes-access-gateway/pull/29/files#diff-786059d243e2e03a9358891e734671f3d86a85cfa09464b8b785ff1ca9735e43R215) that updates this `lastLogTime` at the end of each command.